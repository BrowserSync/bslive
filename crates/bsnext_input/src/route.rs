use std::collections::BTreeMap;
use std::fmt::{Display, Formatter};
use std::hash::{Hash, Hasher};
use std::ops::Deref;

use crate::watch_opts::WatchOpts;
use bsnext_resp::inject_opts::InjectOpts;

#[derive(Debug, PartialEq, Hash, Clone, serde::Deserialize, serde::Serialize)]
pub struct Route {
    pub path: String,
    #[serde(flatten)]
    pub kind: RouteKind,
    #[serde(flatten)]
    pub opts: Opts,
}

#[derive(Debug, Default, PartialEq, Hash, Clone, serde::Deserialize, serde::Serialize)]
pub struct Opts {
    #[serde(flatten)]
    pub cors: Option<CorsOpts>,
    #[serde(flatten)]
    pub delay: Option<DelayOpts>,
    #[serde(default)]
    pub watch: WatchOpts,
    #[serde(default)]
    pub inject: InjectOpts,
    pub headers: Option<BTreeMap<String, String>>,
}

impl Default for Route {
    fn default() -> Self {
        Self {
            path: "/".to_string(),
            kind: RouteKind::new_html("default"),
            opts: Opts {
                headers: None,
                cors: None,
                delay: None,
                watch: Default::default(),
                inject: Default::default(),
            },
        }
    }
}

impl AsRef<Route> for Route {
    fn as_ref(&self) -> &Route {
        self
    }
}

impl Route {
    pub fn path(&self) -> &str {
        self.path.as_str()
    }
}

#[derive(Debug, Hash, PartialEq, Clone, serde::Deserialize, serde::Serialize)]
#[serde(untagged)]
pub enum RouteKind {
    Raw(RawRoute),
    Proxy(ProxyRoute),
    Dir(DirRoute),
}

impl RouteKind {
    pub fn new_html(html: impl Into<String>) -> Self {
        RouteKind::Raw(RawRoute::Html { html: html.into() })
    }
    pub fn new_raw(raw: impl Into<String>) -> Self {
        RouteKind::Raw(RawRoute::Raw { raw: raw.into() })
    }
    pub fn new_json(json: JsonWrapper) -> Self {
        RouteKind::Raw(RawRoute::Json { json })
    }
    pub fn new_sse(raw: impl Into<String>) -> Self {
        RouteKind::Raw(RawRoute::Sse { sse: raw.into() })
    }
}

#[derive(Debug, PartialEq, Clone, serde::Deserialize, serde::Serialize)]
pub struct JsonWrapper(pub serde_json::Value);

impl Display for JsonWrapper {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", serde_json::to_string(&self.0).expect("serde_json"))
    }
}

impl Deref for JsonWrapper {
    type Target = serde_json::Value;

    fn deref(&self) -> &Self::Target {
        &self.0
    }
}

impl Hash for JsonWrapper {
    fn hash<H: Hasher>(&self, state: &mut H) {
        //todo: implement hashing for serde_value
        if let Ok(as_str) = serde_json::to_string(&self.0) {
            state.write(as_str.as_bytes());
        } else {
            todo!("handle error here?")
        }
    }
}

#[derive(Debug, PartialEq, Hash, Clone, serde::Deserialize, serde::Serialize)]
pub struct DirRoute {
    pub dir: String,
}

#[derive(Debug, PartialEq, Hash, Clone, serde::Deserialize, serde::Serialize)]
pub struct ProxyRoute {
    pub proxy: String,
}

#[derive(Debug, PartialEq, Hash, Clone, serde::Deserialize, serde::Serialize)]
#[serde(untagged)]
pub enum RawRoute {
    Html { html: String },
    Json { json: JsonWrapper },
    Raw { raw: String },
    Sse { sse: String },
}

#[derive(Debug, PartialEq, Hash, Clone, serde::Deserialize, serde::Serialize)]
pub enum CorsOpts {
    #[serde(rename = "cors")]
    Cors(bool),
}

#[derive(Debug, PartialEq, Hash, Clone, serde::Deserialize, serde::Serialize)]
pub enum DelayOpts {
    #[serde(rename = "delay")]
    Delay(DelayKind),
}

#[derive(
    Debug, Ord, PartialOrd, PartialEq, Eq, Hash, Clone, serde::Deserialize, serde::Serialize,
)]
pub enum DelayKind {
    #[serde(rename = "ms")]
    Ms(u64),
}

#[derive(
    Debug, Ord, PartialOrd, PartialEq, Eq, Hash, Clone, serde::Deserialize, serde::Serialize,
)]
pub enum DebounceDuration {
    #[serde(rename = "ms")]
    Ms(u64),
}

#[derive(
    Debug, Ord, PartialOrd, PartialEq, Eq, Hash, Clone, serde::Deserialize, serde::Serialize,
)]
#[serde(untagged)]
pub enum FilterKind {
    StringGlob(String),
    Extension { ext: String },
    Glob { glob: String },
    List(Vec<FilterKind>),
}

#[derive(
    Debug, Ord, PartialOrd, PartialEq, Eq, Hash, Clone, serde::Deserialize, serde::Serialize,
)]
pub struct Spec {
    #[serde(flatten)]
    pub opts: Option<SpecOpts>,
}

#[derive(
    Debug, Ord, PartialOrd, PartialEq, Eq, Hash, Clone, serde::Deserialize, serde::Serialize,
)]
pub struct SpecOpts {
    pub debounce: Option<DebounceDuration>,
    pub filter: Option<FilterKind>,
}

#[derive(
    Debug, PartialOrd, Ord, Eq, PartialEq, Hash, Clone, serde::Deserialize, serde::Serialize,
)]
pub struct Watcher {
    pub dir: String,
    pub debounce_ms: Option<u64>,
    pub filter: Option<FilterKind>,
}
