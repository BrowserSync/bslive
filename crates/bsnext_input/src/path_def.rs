use crate::route::PathDefError;
use serde::{de, Deserialize, Deserializer, Serializer};
use std::path::{Component, Path, PathBuf};
use std::str::FromStr;

#[derive(Debug, PartialEq, Hash, Clone)]
pub struct PathDef {
    inner: String,
}

impl FromStr for PathDef {
    type Err = PathDefError;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        Self::try_new(s)
    }
}

pub fn deserialize<'de, D>(deserializer: D) -> Result<PathDef, D::Error>
where
    D: Deserializer<'de>,
{
    let s = String::deserialize(deserializer)?;
    FromStr::from_str(&s).map_err(de::Error::custom)
}

pub fn serialize<S>(x: &PathDef, s: S) -> Result<S::Ok, S::Error>
where
    S: Serializer,
{
    s.serialize_str(x.inner.as_str())
}

impl PathDef {
    pub fn as_pb(&self) -> PathBuf {
        PathBuf::from(&self.inner)
    }
    pub fn as_str(&self) -> &str {
        self.inner.as_str()
    }
    pub fn try_new<A: AsRef<str>>(input: A) -> Result<Self, PathDefError> {
        let str = input.as_ref();
        let is_path = str.starts_with("/");
        let p = Path::new(str);
        let has_star = p.components().any(|c| match c {
            Component::Prefix(_) => false,
            Component::RootDir => false,
            Component::CurDir => false,
            Component::ParentDir => false,
            Component::Normal(n) => n.to_str().is_some_and(|str| str.contains("*")),
        });
        let route = match (is_path, has_star) {
            (true, false) => Ok(Self {
                inner: str.to_string(),
            }),
            (true, true) => Err(PathDefError::ContainsStar),
            (false, _) => Err(PathDefError::StartsWithSlash),
        }?;

        let mut r = matchit::Router::new();
        match r.insert(&route.inner, true) {
            Ok(_) => Ok(route),
            Err(e) => Err(PathDefError::InsertError(e)),
        }
    }
}

#[cfg(test)]
mod test {
    use super::*;
    #[test]
    fn test_verify() -> anyhow::Result<()> {
        let input = "/abc/one/*rest";
        let actual = PathDef::try_new(input).unwrap_err();
        assert_eq!(actual, PathDefError::ContainsStar);

        let input = "abc/one";
        let actual = PathDef::try_new(input).unwrap_err();
        assert_eq!(actual, PathDefError::StartsWithSlash);

        let input = "/:";
        let actual = PathDef::try_new(input).unwrap_err();
        assert_eq!(
            actual,
            PathDefError::InsertError(matchit::InsertError::UnnamedParam)
        );

        let input = "/:abc:abc";
        let actual = PathDef::try_new(input).unwrap_err();
        assert_eq!(
            actual,
            PathDefError::InsertError(matchit::InsertError::TooManyParams)
        );

        Ok(())
    }
}
