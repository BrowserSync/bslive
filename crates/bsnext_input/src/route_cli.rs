use crate::path_def::PathDef;
use crate::route::{CorsOpts, DelayKind, DelayOpts, DirRoute, Opts, ProxyRoute, Route, RouteKind};
use clap::Parser;
use shell_words::split;

#[derive(clap::Parser, Debug)]
#[command(version)]
pub struct RouteCli {
    #[command(subcommand)]
    command: SubCommands,
}

impl RouteCli {
    pub fn try_from_cli_str<A: AsRef<str>>(a: A) -> Result<RouteCli, anyhow::Error> {
        let as_args = split(a.as_ref())?;
        RouteCli::try_parse_from(as_args).map_err(|e| anyhow::anyhow!(e))
    }
}

impl TryInto<Route> for RouteCli {
    type Error = anyhow::Error;

    fn try_into(self) -> Result<Route, Self::Error> {
        Ok(match self.command {
            SubCommands::ServeDir {
                dir, path, opts, ..
            } => Route {
                path: PathDef::try_new(path)?,
                kind: RouteKind::Dir(DirRoute { dir, base: None }),
                opts: opts_to_route_opts(&opts),
                ..std::default::Default::default()
            },
            SubCommands::Proxy { path, target, opts } => Route {
                path: PathDef::try_new(path)?,
                kind: RouteKind::Proxy(ProxyRoute {
                    proxy: target,
                    proxy_headers: None,
                    rewrite_uri: None,
                }),
                opts: opts_to_route_opts(&opts),
                ..std::default::Default::default()
            },
        })
    }
}

fn opts_to_route_opts(cli_opts: &RouteCliSharedOpts) -> Opts {
    Opts {
        delay: cli_opts.delay.map(|ms| DelayOpts::Delay(DelayKind::Ms(ms))),
        cors: cli_opts.cors.map(CorsOpts::Cors),
        ..std::default::Default::default()
    }
}

#[derive(Debug, clap::Subcommand)]
pub enum SubCommands {
    /// does testing things
    Proxy {
        #[arg(long, default_value = "/proxy")]
        path: String,
        #[arg(long)]
        target: String,
        #[clap(flatten)]
        opts: RouteCliSharedOpts,
    },
    ServeDir {
        /// Which path should this directory be served from
        #[arg(long, default_value = "/")]
        path: String,
        /// Which directory should be served
        #[arg(long, default_value = ".")]
        dir: String,
        #[clap(flatten)]
        opts: RouteCliSharedOpts,
    },
}

#[derive(Debug, clap::Parser)]
pub struct RouteCliSharedOpts {
    /// should a ms delay be applied?
    #[arg(long)]
    delay: Option<u64>,
    /// should cors headers be added to responses
    #[arg(long)]
    cors: Option<bool>,
}

#[cfg(test)]
mod test {
    use super::*;
    use clap::Parser;

    use shell_words::split;
    #[test]
    fn test_serve_dir() -> anyhow::Result<()> {
        let input = "bslive serve-dir --path=/ --dir=examples/basic/public";
        let as_args = split(input)?;
        let parsed = RouteCli::try_parse_from(as_args)?;
        let as_route: Route = parsed.try_into().unwrap();
        insta::assert_debug_snapshot!(as_route);
        Ok(())
    }
    #[test]
    fn test_serve_dir_delay() -> anyhow::Result<()> {
        let input = "bslive serve-dir --path=/ --dir=examples/basic/public --delay=1000";
        let as_args = split(input)?;
        let parsed = RouteCli::try_parse_from(as_args)?;
        let as_route: Route = parsed.try_into().unwrap();
        insta::assert_debug_snapshot!(as_route);
        Ok(())
    }
    #[test]
    fn test_serve_dir_shared() -> anyhow::Result<()> {
        let input =
            "bslive serve-dir --path=/ --dir=examples/basic/public --delay=1000 --cors=true";
        let as_args = split(input)?;
        let parsed = RouteCli::try_parse_from(as_args)?;
        let as_route: Route = parsed.try_into().unwrap();
        insta::assert_debug_snapshot!(as_route);
        Ok(())
    }
}
