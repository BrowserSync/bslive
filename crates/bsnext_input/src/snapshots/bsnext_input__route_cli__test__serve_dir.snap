---
source: crates/bsnext_input/src/route_cli.rs
expression: as_route
---
Route {
    path: PathDef {
        inner: "/",
    },
    kind: Dir(
        DirRoute {
            dir: "examples/basic/public",
            base: None,
        },
    ),
    opts: Opts {
        cors: None,
        delay: None,
        watch: Bool(
            true,
        ),
        inject: Bool(
            true,
        ),
        headers: None,
        cache: Prevent,
        compression: None,
    },
    fallback: None,
    when: None,
}
